junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java

```diff
@@ -117,23 +117,32 @@ public void registerTestExecutionListeners(TestExecutionListener... listeners) {
 	@Override
 	public TestPlan discover(LauncherDiscoveryRequest discoveryRequest) {
 		Preconditions.notNull(discoveryRequest, "LauncherDiscoveryRequest must not be null");
-		return TestPlan.from(discoverRoot(discoveryRequest, "discovery").getEngineDescriptors());
+		return UnmodifiableTestPlan.from(discoverRoot(discoveryRequest, "discovery"));
 	}
 
 	@Override
 	public void execute(LauncherDiscoveryRequest discoveryRequest, TestExecutionListener... listeners) {
 		Preconditions.notNull(discoveryRequest, "LauncherDiscoveryRequest must not be null");
 		Preconditions.notNull(listeners, "TestExecutionListener array must not be null");
 		Preconditions.containsNoNullElements(listeners, "individual listeners must not be null");
-		execute(discoverRoot(discoveryRequest, "execution"), discoveryRequest.getConfigurationParameters(), listeners);
+		execute(UnmodifiableTestPlan.from(discoverRoot(discoveryRequest, "execution")), listeners);
+	}
+
+	@Override
+	public void execute(TestPlan testPlan, TestExecutionListener... listeners) {
+		Preconditions.notNull(testPlan, "TestPlan must not be null");
+		Preconditions.condition(testPlan instanceof UnmodifiableTestPlan, "TestPlan was not returned by this Launcher");
```

Comment:
Did you consider to check the identity of the test plan somehow (e.g. by some fingerprint)?
I guess a modified UnmodifiableTestPlan could somehow be engineered, maybe by serialization tricks...